src/login/logind.c

```diff
@@ -444,74 +444,80 @@ static int parse_fdname(const char *fdname, char **session_id, dev_t *dev) {
 
-                fd = SD_LISTEN_FDS_START + i;
+        if (!S_ISCHR(st.st_mode) || st.st_rdev != dev) {
+                log_debug("Device fd doesn't point to the expected character device node");
+                goto close;
+        }
 
-                r = parse_fdname(fdnames[i], &id, &dev);
-                if (r < 0) {
-                        log_debug_errno(r, "Failed to parse fd name %s: %m", fdnames[i]);
-                        close_nointr(fd);
-                        continue;
-                }
+        sd = hashmap_get(s->devices, &dev);
+        if (!sd) {
+                /* Weird, we got an fd for a session device which wasn't
+                 * recorded in the session state file... */
+                log_warning("Got fd for missing session device [%u:%u] in session %s",
+                            major(dev), minor(dev), s->id);
+                goto close;
+        }
 
-                s = hashmap_get(m->sessions, id);
-                if (!s) {
-                        /* If the session doesn't exist anymore, the associated session
-                         * device attached to this fd doesn't either. Let's simply close
-                         * this fd. */
-                        log_debug("Failed to attach fd for unknown session: %s", id);
-                        close_nointr(fd);
-                        continue;
-                }
+        log_debug("Attaching fd to session device [%u:%u] for session %s",
+                  major(dev), minor(dev), s->id);
 
-                if (fstat(fd, &st) < 0) {
-                        /* The device is allowed to go away at a random point, in which
-                         * case fstat failing is expected. */
-                        log_debug_errno(errno, "Failed to stat device fd for session %s: %m", id);
-                        close_nointr(fd);
-                        continue;
-                }
+        session_device_attach_fd(sd, fd, s->was_active);
+        return;
 
-                if (!S_ISCHR(st.st_mode) || st.st_rdev != dev) {
-                        log_debug("Device fd doesn't point to the expected character device node");
-                        close_nointr(fd);
-                        continue;
-                }
+close:
+        r = asprintf(&message, "FDSTOREREMOVE=1\n"
+                               "FDNAME=%s\n",
+                               fdname);
+        if (r >= 0)
+                (void) sd_notify(false, message);
 
-                sd = hashmap_get(s->devices, &dev);
-                if (!sd) {
-                        /* Weird, we got an fd for a session device which wasn't
-                         * recorded in the session state file... */
-                        log_warning("Got fd for missing session device [%u:%u] in session %s",
-                                    major(dev), minor(dev), s->id);
-                        close_nointr(fd);
-                        continue;
-                }
+        close_nointr(fd);
```

Comment:
maybe change the parameter name to _fd, and then declare `_cleanup_close int fd = _fd;` in the first line of this function, and then explicitly unset fd = -1 if we actually use it. That way you can avoid all the goto stuff, and the error path is automatic when you just return on each check, and the code will be much shorter.

Reply:
The goto stuff is also used to send the FDSTOREREMOVE message.  (I will hint again, this is clearer if you use the `?w=1` url trick to review without indentation changes :-P.  Or tell me to split into two commits: refactor first and then the  improvement).

So far I'm happy without the fd variable dance.  Unless we have some macro TAKE() that I haven't noticed, in the sense of [Option::take()](https://doc.rust-lang.org/std/option/enum.Option.html#method.take)?

I can see one other way, but I thought I preferred this one, where this one function takes complete ownership of the fd.

Reply:
ah, right, regarding the goto, ignore me

Reply:
A TAKE() macro might be sexy to have though. We currently have free_and_replace() which does something similar (though for dynamic memory). Maybe we should introduce a set of macros for that purpose, that also handle other types of resources.


Maybe:

```c
#define TAKE_FD(x) ({ int __x = (x); (x) = -1; __x; })
#define TAKE_MEMORY(x) ({ typeof(x) __x = (x); (x) = NULL; __x; })
```

Example:

```c
…
fd = TAKE_FD(old_fd);
assert(old_fd < 0);

ptr = TAKE_MEMORY(old_ptr);
assert(!old_ptr);
…
```

Reply:
I implemented your suggestion now, see #8553, I think it's kinda nice. Thank you for the suggestion! PTAL!